### PR TITLE
Sync test files to GitHub

### PR DESCRIPTION
**Pull request by flaurida**
* Part 1 of #1202
* Allow GitHub sync integration on team
* List branches in dashboard
* Choosing a branch shows tests on that branch
* Sync creating and deleting tests with GitHub repo

### CODE REVIEW

File: app/.env.example [hunk 1]
```diff
@@ -31,13 +31,14 @@ EMAIL_DOMAIN=dev.qawolf.email
...
 
-GITHUB_APP_CLIENT_ID=client-id
 GITHUB_APP_CLIENT_SECRET=secret
 GITHUB_APP_ID=11
 GITHUB_OAUTH_CLIENT_ID=OAUTH_CLIENT_ID
 GITHUB_OAUTH_CLIENT_SECRET=OAUTH_CLIENT_SECRET
+GITHUB_SYNC_APP_CLIENT_SECRET=secret
 GITHUB_USER_AGENT=Github App Name
 GITHUB_WEBHOOK_SECRET=secret
+GITHUB_SYNC_APP_ID=12
```
Comment (jperl) [inline app/.env.example:41]: abc

**Comment by jperl** [comment]
I installed it to an empty github repo and got an internal server error

```
critical unexpected error Git Repository is empty. INTERNAL_SERVER_ERROR [GraphQLError [Object]: Git Repository is empty.] {
  locations: [ { line: 2, column: 3 } ],
  path: [ 'tests' ],
  extensions: {
    code: 'INTERNAL_SERVER_ERROR',
    exception: {
      name: 'HttpError',
      status: 409,
      headers: [Object],
      request: [Object],
      documentation_url: 'https://docs.github.com/rest/reference/git#get-a-tree',
      stacktrace: [Array]
    }
  }
}
```

Should we catch this and provide it as a client error or na?

File: app/components/GitHubIntegration.tsx [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+import Spinner from "./shared/Spinner";
+import Text from "./shared/Text";
+import { StateContext } from "./StateContext";
+
+type Props = { isSync?: boolean };
+
+export default function GitHubIntegration({ isSync }: Props): JSX.Element {
+  useEnsureUser();
+
+  console.log("SYNC", isSync);
```
Comment (jperl) [inline app/components/GitHubIntegration.tsx:20]: straggler

File: app/hooks/apiKey.ts [hunk 3]
```diff
@@ -1,22 +1,11 @@
...
-export const useApiKey = (): UseApiKey => {
-  const { teamId } = useContext(StateContext);
-
-  const { data } = useTeam({ id: teamId });
-  const team = data?.team || null;
-
-  const apiKey = team?.api_key || copy.apiKeyHere;
+export const useApiKey = (): string => {
+  const { team } = useContext(UserContext);
+  console.log("TEAM", team);
```
Comment (jperl) [inline app/hooks/apiKey.ts:8]: straggler

File: app/server/resolvers/test/helpers.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+  const fileWithoutFolder = last(file.split("/"));
+
+  return fileWithoutFolder;
+};
+
+const createMissingTests = async (
+  { gitHubTests, team_id, tests }: CreateMissingTests,
+  options: ModelOptions
+): Promise<Test[]> => {
+  const missingTests = gitHubTests.filter((test) => {
```
Comment (jperl) [inline app/server/resolvers/test/helpers.ts:40]: We need to think through a better matching logic that supports folders IMO. Perhaps we should store the path

File: app/server/resolvers/test/index.ts [hunk 5]
```diff
@@ -64,6 +67,15 @@ export const createTestResolver = async (
     { db, logger }
   );
 
+  if (branch && team.git_sync_integration_id) {
```
Comment (jperl) [inline app/server/resolvers/test/index.ts:70]: This should throw an error if !team.git_sync_integration_id

File: app/server/migrations/20210424144516_update_tests_name_nullable.ts [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+export async function up(knex: Knex): Promise<void> {
+  await knex.schema.alterTable("tests", (table) => {
+    table.string("name").nullable().alter();
+  });
+
+  await knex.raw(
+    `ALTER TABLE tests ADD CONSTRAINT tests_require_name_or_path CHECK (num_nonnulls(name, path) = 1)`
+  );
+
+  return knex.raw(`CREATE UNIQUE INDEX tests_unique_path_team_id
```
Comment (jperl) [inline app/server/migrations/20210424144516_update_tests_name_nullable.ts:12]: Will this be a problem with nulls?

Comment (flaurida) [reply app/server/migrations/20210424144516_update_tests_name_nullable.ts:12]: i'll check

File: app/server/resolvers/test/helpers.ts [hunk 7]
```diff
@@ -0,0 +1,119 @@
+import { camelCase, last, uniq } from "lodash";
```
Comment (jperl) [inline app/server/resolvers/test/helpers.ts:1]: perhaps this file should be named `gitHubTests`?

File: app/server/migrations/20210424144516_update_tests_name_nullable.ts (hunk 6 above)
Comment (flaurida) [reply app/server/migrations/20210424144516_update_tests_name_nullable.ts:12]: no

File: app/server/resolvers/test/index.ts (hunk 5 above)
Comment (jperl) [inline app/server/resolvers/test/index.ts:70]: Also I think this should all be in a transaction so the test insertion would rollback if the commit fails.

File: app/server/models/test.ts [hunk 8]
```diff
@@ -278,6 +262,43 @@ export const findTestForRun = async (
   return test;
 };
 
+export const findTestsForTeam = async (
```
Comment (jperl) [inline app/server/models/test.ts:265]: abc

File: app/server/resolvers/test/index.ts [hunk 9]
```diff
@@ -53,35 +56,65 @@ export const createTestResolver = async (
...
+  { branch, ids }: DeleteTestsMutation,
   { db, logger, teams }: Context
 ): Promise<Test[]> => {
+  const log = logger.prefix("deleteTestsResolver");
+  log.debug("ids", ids);
+
+  if (branch) {
+    log.debug("delete from git branch", branch);
+
+    const tests = await findTests(ids, { db, logger });
```
Comment (jperl) [inline app/server/resolvers/test/index.ts:98]: Seems like this auth logic could be done once

File: app/server/resolvers/test/gitHubTests.ts [hunk 10]
```diff
@@ -0,0 +1,100 @@
+import { uniq } from "lodash";
```
Comment (jperl) [inline app/server/resolvers/test/gitHubTests.ts:1]: Should this file actually be in services or models 🤔 ?

Comment (flaurida) [reply app/server/resolvers/test/gitHubTests.ts:1]: u move it

**Pull request merged**